Page about granting access to projects and namespaces (#105)
appuio/appuio-cloud-docs: docs/modules/ROOT/pages/how-to/grant-access.adoc

= Granting Access to Projects and Namespaces

In the xref:how-to/manage-projects-and-namespaces.adoc[Managing Projects and Namespaces] page it's explained how to create projects and namespaces.
This page explains in detail how to manage users, groups, and teams within those projects, assigning rights and entitlements.

> IMPORTANT: The contents of this page rely on knowledge provided by the https://docs.openshift.com/container-platform/4.9/authentication/using-rbac.html[official OpenShift 4.9 documentation about Role-Base Access Control (RBAC)], highlighting the differences in {product}.

Reviewer: ```suggestion
IMPORTANT: The contents of this page rely on knowledge provided by the https://docs.openshift.com/container-platform/latest/authentication/using-rbac.html[official OpenShift documentation about Role-Base Access Control (RBAC)], highlighting the differences in {product}.
```

== Organizations and Teams in {product}

> Organizations and teams are reflected as OpenShift 4.9 groups.

Reviewer 2: ```suggestion
Organizations and teams are reflected as OpenShift groups.
```

For each and every organization, there is a group with the same name; and for each and every team, there is a group with the same name, but prefixed with the name of the parent organization.
For example: `acme+accounting` would be a valid team name in {product} for the Accounting team within the ACME organization.

IMPORTANT: OpenShift groups don't have any notion of hierarchy! From the point of view of OpenShift, they all belong to the same flat structure.

== Granting Privileges

> A `RoleBinding` object grants the creator of a Project or Namespace admin privileges.

Reviewer 2: That one isn't clear. Which `RoleBinding`?

> A `RoleBinding` object grants the creator of a Project or Namespace admin privileges.

Reviewer: ```suggestion
The RoleBinding object named `admin` grants the creator of a Project or Namespace admin privileges.
```

> This is a default behavior in OpenShift.

Reviewer 2: That's not true IMHO? Or what does "This" refer to?

> This means that by default, all members of the owning organization are granted admin privileges by referencing the corresponding OpenShift Group.

Reviewer 2: IMHO this is APPUiO Cloud specific?

> This means that by default, all members of the owning organization are granted admin privileges by referencing the corresponding OpenShift Group.

Reviewer: ```suggestion
In addition, the RoleBinding named `namespace-owner` grants the group named equal to the Organization owning the namespace admin privileges.
This means that by default, all members of the owning organization are granted admin privileges.
```

> Both RoleBindings can be deleted and replaced by something else.

Reviewer 2: Which "both"? How are they called?

Reviewer: Should be addressed by my suggestions.

> Both RoleBindings can be deleted and replaced by something else.

Reviewer: ```suggestion
Both RoleBindings can be changed or deleted and replaced by something else.
```

CAUTION: Modifications to RoleBindings can prevent users from accessing their projects; in other words, you might end up locked out of your project!

> Access can be granted to any OpenShift Group (This includes that namespace owned by organization A can be managed by members of organization B or only a team of organization B)

Reviewer: ```suggestion
Access can be granted to any OpenShift Group.
This includes that namespace owned by organization A can be managed by members of organization B or only a team of organization B.
```

= Granting Access to Projects and Namespaces

In the xref:how-to/manage-projects-and-namespaces.adoc[Managing Projects and Namespaces] page it's explained how to create projects and namespaces.
This page explains in detail how to manage users, groups, and teams within those projects, assigning rights and entitlements.

IMPORTANT: The contents of this page rely on knowledge provided by the https://docs.openshift.com/container-platform/latest/authentication/using-rbac.html[official OpenShift documentation about Role-Base Access Control (RBAC)], highlighting the differences in {product}.

== Organizations and Teams in {product}

Organizations and teams are reflected as OpenShift groups.
For each and every organization, there is a group with the same name; and for each and every team, there is a group with the same name, but prefixed with the name of the parent organization.
For example: `acme+accounting` would be a valid team name in {product} for the Accounting team within the ACME organization.

IMPORTANT: OpenShift groups don't have any notion of hierarchy! From the point of view of OpenShift, they all belong to the same flat structure.

== Granting Privileges

> The RoleBinding object named `admin` grants the creator of a Project or Namespace admin privileges.
> This is a default behavior in OpenShift.
> 
> In addition, the RoleBinding named `namespace-owner` grants the group named equal to the Organization owning the namespace admin privileges.
> This means that by default, all members of the owning organization are granted admin privileges.

Reviewer 3: These paragraphs are not correct.

1. We patch the `admin` RoleBinding to grant the organization admin privileges on the project, cf. https://hub.syn.tools/appuio-cloud/references/policies/10_generate_default_rolebinding_in_ns.html
2. We create the `namespace-owner` RoleBinding, which allows the organization to view, edit and delete the Namespace object itself. Edit restrictions apply, cf. https://hub.syn.tools/appuio-cloud/references/policies/02_validate_namespace_metadata.html

Maybe rewrite the paragraphs along the lines of the suggestion below (I'm not happy  with the wording, feel free to improve it).

```suggestion
On {product}, the RoleBinding object named `admin` grants the group named equal to the Organization owning the Project or Namespace admin privileges.
This differs from the default behavior in OpenShift, where the default `admin` RoleBinding only grants the creator of a Project admin privileges.

In addition, the RoleBinding named `namespace-owner` gives the group named equal to the Organization of the Namespace object permissions to view, edit and delete the object.
```
